cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -350,6 +373,33 @@ public Void call() throws Exception {
   }
 
   /**
+   * Calls the destroy method of {@link ProgramLifecycle}.
+   */
+  private void destroy(final ProgramState state, final BasicSparkClientContext context) {
+    DynamicDatasetCache datasetCache = runtimeContext.getDatasetCache();
+    TransactionContext txContext = datasetCache.newTransactionContext();
+    try {
+      Transactions.execute(txContext, spark.getClass().getName() + ".destroy()", new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          Cancellable cancellable = SparkRuntimeUtils.setContextClassLoader(new SparkClassLoader(runtimeContext));
+          try {
+            context.setState(state);
```

Comment:
I prefer the state set outside of this method, probably right after the `jobCompletion.get()` in the `run()` method. Otherwise for program that are still using the old onFinish() method, if they call `context.getState()`, it will be wrong).